swim/heal_via_discover_provider.go

```diff
@@ -0,0 +1,166 @@
+// the discoverProvider for nodes that are faulty or not available in its
+// membership. It attempts the heal probabalisticly so that the discovery provider
+// doesn't get overloaded with request -- with default settings 6
+// times per minutes on avarage for the entire cluster.
+type discoverProviderHealer struct {
+	node *Node
+
+	discoverProvider *discovery.DiscoverProvider
+	baseProbabillity float64
+	period           time.Duration
+
+	previousHostListSize int
+	quit                 chan struct{}
+
+	logger log.Logger
+}
+
+func newDiscoverProviderHealer(n *Node, baseProbability float64, period time.Duration) *discoverProviderHealer {
+	return &discoverProviderHealer{
+		node:             n,
+		baseProbabillity: baseProbability,
+		period:           period,
+		quit:             make(chan struct{}),
+		logger:           logging.Logger("healer").WithField("local", n.Address()),
+	}
+}
+
+// Start the partition healing loop
+func (h *discoverProviderHealer) Start() {
+	go func() {
+		for {
+			// attempt heal with the pro
+			if rand.Float64() < h.Probability() {
+				h.Heal()
+			}
+
+			// loop or quit
+			select {
+			case <-h.node.clock.After(h.period):
+			case <-h.quit:
+				return
+			}
+		}
+	}()
+}
+
+// Stop the partition healing loop.
+func (h *discoverProviderHealer) Stop() {
+	select {
+	case <-h.quit:
+	default:
+		close(h.quit)
+	}
+}
+
+// Probability returns the probability when a heal should be attempted
+// we want to throttle the heal attempts to alleviate pressure on the
+// discover provider.
+func (h *discoverProviderHealer) Probability() float64 {
+	// avoid division by zero.
+	if h.previousHostListSize < h.node.CountReachableMembers() {
+		h.previousHostListSize = h.node.CountReachableMembers()
+	}
+	if h.previousHostListSize < 1 {
+		h.previousHostListSize = 1
+	}
+	return h.baseProbabillity / float64(h.previousHostListSize)
+}
+
+// Heal iterates over the hostList that the discoverProvider provides. If the
+// node encounters a host that is faulty or not in the membership, we pick that
+// node as a target to perform a partition heal with.
+//
+// If heal was attempted, returns identities of the target nodes.
+func (h *discoverProviderHealer) Heal() []string {
+	h.node.emit(DiscoHealEvent{})
+	// get list from discovery provider
+	if h.node.discoverProvider == nil {
+		return []string{}
+	}
+	hostList, err := h.node.discoverProvider.Hosts()
+	if err != nil {
+		h.logger.Warn("unable to receive host list from discover provider")
+		return []string{}
+	}
+
+	h.previousHostListSize = len(hostList)
+
+	var targets []string
+	for _, address := range hostList {
+		m, ok := h.node.memberlist.Member(address)
+		if !ok || m.Status == Faulty {
+			targets = append(targets, address)
+		}
+	}
+
+	// filter hosts that we already know about and attempt to heal nodes that
+	// are complementary to the membership of this node.
+	var ret []string
+
+	for len(targets) != 0 {
```

Comment:
I haven't reviewed this diff, so my response is generic.

IMHO, there ought to be an independent process which considers the list of expected, but unavailable peers (EBUPs), and pingjoins a small constant number of them per period (1-3). That period ought to be reasonable but short (1-5s?); in the normal/stable case, it will trigger nothing.

Possible tweaks: When a pingjoin initiator adds (a) new node(s), it could also immediately enqueue another pingjoin op, with the ideal that healing begets healing. It could equally retard that timer when all EBUPs have repeatedly timed out, although this seems less useful if the main timer is well set. IIC, this lets many small partitions seek each other out less frequently, while allowing the first healers to accelerate upon contact.

@Motiejus I don't wholly buy the notion that "buildup" syncs are super expensive, or that bandwidth caps are a threat. Do we have measurements of a full-sync when a singleton joins a 1k cluster? I would be curious to know the numbers.